data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala

```diff
@@ -0,0 +1,200 @@
+                          batchId: String,
+                          contentId: String)(implicit config: JobConfig): Unit = {
+    val certIdsDF: DataFrame = incorrectRecords.select("course_id", "batch_id", "content_id", "user_id").distinct()
+      .join(userEnrolmentDF, incorrectRecords.col("user_id") === userEnrolmentDF.col("userid") && incorrectRecords.col("course_id") === userEnrolmentDF.col("courseid") &&
+        incorrectRecords.col("batch_id") === userEnrolmentDF.col("batchid"), "left_outer")
+      .withColumn("certificate_data", explode_outer(col("issued_certificates")))
+      .withColumn("certificate_id", col("certificate_data.identifier"))
+      .select("courseid", "batchid", "userid", "certificate_id").filter(col("certificate_id") =!= "")
+    saveLocal(certIdsDF, batchId, contentId = contentId, "revoked-cert-data")
+  }
+
+  def saveLocal(data: DataFrame,
+                batchId: String,
+                contentId: String,
+                folderName: String)(implicit config: JobConfig): Unit = {
+    JobLogger.log("Generating the CSV File", Option(Map("batch_id" -> batchId, "content_id" -> contentId)), INFO)
+    val outputPath = config.modelParams.getOrElse(Map[String, Option[AnyRef]]()).getOrElse("csvPath", "").asInstanceOf[String]
+    data.repartition(1).write.option("header", value = true).format("com.databricks.spark.csv").save(outputPath.concat(s"/$folderName-$batchId-$contentId-${System.currentTimeMillis()}.csv"))
+  }
+
+
+  def removeAssessmentRecords(incorrectRecords: DataFrame,
+                              batchId: String,
+                              contentId: String,
+                              isDryRunMode: Boolean)(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig, sc: SparkContext): Unit = {
+    if (isDryRunMode) {
+      saveLocal(incorrectRecords, batchId, contentId, "assessment-invalid-attempts-records")
+    } else {
+      JobLogger.log("Deleting the records from the table", None, INFO)
+      saveLocal(incorrectRecords, batchId, contentId, "assessment-invalid-attempts-records") // For cross verification purpose
+      incorrectRecords.select("course_id", "batch_id", "user_id", "content_id", "attempt_id").rdd.deleteFromCassandra(AppConf.getConfig("sunbird.courses.keyspace"), "assessment_aggregator", keyColumns = SomeColumns("course_id", "batch_id", "user_id", "content_id", "attempt_id"))
+    }
+  }
+
+  def generateInstructionEvents(inCorrectRecordsDF: DataFrame,
+                                batchId: String,
+                                contentId: String)(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig, sc: SparkContext): Unit = {
+    val outputPath = config.modelParams.getOrElse(Map[String, Option[AnyRef]]()).getOrElse("csvPath", "").asInstanceOf[String]
+    val file = new File(outputPath.concat(s"/instruction-events-$batchId-$contentId-${System.currentTimeMillis()}.json"))
+    var writer: BufferedWriter = null
```

Comment:
The BufferedWriter can be created outside of the try-catch-finally block. Also, each line need not be flushed to the file since we are already using a BufferedWriter.